code/harness/DiskContents.cpp

```diff
@@ -0,0 +1,298 @@
+  } else if (a == NULL || stat_attr == NULL) {
+    return false;
+  }
+  return ((stat_attr->st_ino == a->st_ino) &&
+    (stat_attr->st_mode == a->st_mode) &&
+    (stat_attr->st_nlink == a->st_nlink) &&
+    (stat_attr->st_uid == a->st_uid) &&
+    (stat_attr->st_gid == a->st_gid) &&
+    // (stat_attr->st_rdev == a->st_rdev) &&
+    // (stat_attr->st_dev == a->st_dev) &&
+    (stat_attr->st_size == a->st_size) &&
+    (stat_attr->st_blksize == a->st_blksize) &&
+    (stat_attr->st_blocks == a->st_blocks));
+}
+
+bool fileAttributes::compare_md5sum(std::string a) {
+  return md5sum.compare(a);
+}
+
+bool fileAttributes::is_regular_file() {
+  return S_ISREG(stat_attr->st_mode);
+}
+
+std::ofstream& operator<< (std::ofstream& os, fileAttributes& a) {
+  // print dir_attr
+  if (a.dir_attr != NULL) {
+    os << "---Directory Atrributes---" << std::endl;
+    os << "Name   : " << (a.dir_attr)->d_name << std::endl;
+    os << "Inode  : " << (a.dir_attr)->d_ino << std::endl;
+    os << "Offset : " << (a.dir_attr)->d_off << std::endl;
+    os << "Length : " << (a.dir_attr)->d_reclen << std::endl;
+    os << "Type   : " << (a.dir_attr)->d_type << std::endl;
+  }
+  // print stat_attr
+  if (a.stat_attr != NULL) {
+    os << "---File Stat Atrributes---" << std::endl;
+    os << "Inode     : " << (a.stat_attr)->st_ino << std::endl;
+    os << "TotalSize : " << (a.stat_attr)->st_size << std::endl;
+    os << "BlockSize : " << (a.stat_attr)->st_blksize << std::endl;
+    os << "#Blocks   : " << (a.stat_attr)->st_blocks << std::endl;
+    os << "#HardLinks: " << (a.stat_attr)->st_nlink << std::endl;
+    os << "Mode      : " << (a.stat_attr)->st_mode << std::endl;
+    os << "User ID   : " << (a.stat_attr)->st_uid << std::endl;
+    os << "Group ID  : " << (a.stat_attr)->st_gid << std::endl;
+    os << "Device ID : " << (a.stat_attr)->st_rdev << std::endl;
+    os << "RootDev ID: " << (a.stat_attr)->st_dev << std::endl;
+  }
+}
+
+DiskContents::DiskContents(char* path, const char* type) {
+  disk_path = (char *) malloc(sizeof(char)*30);
+  mount_point = (char *) malloc(sizeof(char)*40);
+  fs_type = (char *) malloc(sizeof(char)*10);
+  strcpy(disk_path, path);
+  strcpy(fs_type, type);
+  device_mounted = false;
+}
+
+DiskContents::~DiskContents() {
+  // free(disk_path);
+  // free(mount_point);
+  // free(fs_type);
+}
+
+int DiskContents::mount_disk() {
+  // Construct and set mount_point
+  strcpy(mount_point, "/mnt/");
+  strcat(mount_point, (disk_path + 5));
+  // Create the mount directory with read/write/search permissions for owner and group, 
+  // and with read/search permissions for others.
+  int ret = mkdir(mount_point, S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
+  if (ret == -1 && errno != EEXIST) {
+    std::cout << "creating mountpoint failed" << std::endl;
+    return -1;
+  }
+  // Mount the disk
+  if (mount(disk_path, mount_point, fs_type, MS_RDONLY, NULL) < 0) {
+    return -1;
+  }
+  // sleep after mount
+  unsigned int to_sleep = 1;
```

Comment:
You probably don't need to sleep here. The sleep in `c_harness.cpp` was just in case the disk did some stuff after mount that we didn't want to confuse with the user-issued stream of bios